### PR TITLE
Gauss Newton & LM Robustness Fixes

### PR DESCRIPTION
**Pull request by avik-pal**
Merge after #257 

For `GaussNewton` defaults to `QRFactorization(ColumnNorm())`, to ensure we can directly solve for
$$J \delta x = -f(x)$$

For `LM` we solve

$$ \begin{bmatrix}
    J\\
    \lambda D^TD
\end{bmatrix} v = \begin{bmatrix}
    -f(x)\\
    0
\end{bmatrix}  $$

### CODE REVIEW

File: src/gaussnewton.jl [hunk 1]
```diff
@@ -81,15 +81,31 @@ function SciMLBase.__init(prob::NonlinearLeastSquaresProblem{uType, iip}, alg_::
     kwargs...) where {uType, iip}
     alg = get_concrete_algorithm(alg_, prob)
     @unpack f, u0, p = prob
+
+    # Use QR if the user did not specify a linear solver
+    if alg.linsolve === nothing || alg.linsolve isa QRFactorization ||
+       alg.linsolve isa FastQRFactorization
```
Comment (ChrisRackauckas) [inline src/gaussnewton.jl:87]: This doesn't make sense. It should be open so other methods like GMRES apply.

File: src/levenberg.jl [hunk 2]
```diff
@@ -155,8 +163,26 @@ function SciMLBase.__init(prob::Union{NonlinearProblem{uType, iip},
     @unpack f, u0, p = prob
     u = alias_u0 ? u0 : deepcopy(u0)
     fu1 = evaluate_f(prob, u)
-    uf, linsolve, J, fu2, jac_cache, du, JᵀJ, v = jacobian_caches(alg, f, u, p, Val(iip);
-        linsolve_kwargs, linsolve_with_JᵀJ = Val(true))
+
+    # Use QR if the user did not specify a linear solver
```
Comment (ChrisRackauckas) [inline src/levenberg.jl:167]: Iinearsolve will default on non-square

File: src/levenberg.jl [hunk 3]
```diff
@@ -155,8 +163,26 @@ function SciMLBase.__init(prob::Union{NonlinearProblem{uType, iip},
...
+    if (alg.linsolve === nothing || alg.linsolve isa QRFactorization ||
```
Comment (ChrisRackauckas) [inline src/levenberg.jl:168]: missing GMRES, but leave open

File: src/gaussnewton.jl (hunk 1 above)
Comment (avik-pal) [reply src/gaussnewton.jl:87]: GMRES works on non-square? https://github.com/JuliaSmoothOptimizers/Krylov.jl/blob/main/src/gmres.jl

Comment (avik-pal) [reply src/gaussnewton.jl:87]: is there a `can_handle_non_square` trait?

File: src/utils.jl [hunk 4]
```diff
@@ -206,8 +205,75 @@ function __get_concrete_algorithm(alg, prob)
...
+        b = ones(Float64, 3)
+        solve(LinearProblem(A, b), alg)
+        return false
+    catch err
+        return true
+    end
+end
+for alg in (:QRFactorization, :FastQRFactorization)
+    @eval needs_square_A(::$(alg)) = false
+end
```
Comment (avik-pal) [inline src/utils.jl:279]: @ChrisRackauckas is this a good check? :sweat:

File: src/utils.jl [hunk 5]
```diff
@@ -206,8 +205,81 @@ function __get_concrete_algorithm(alg, prob)
...
+_try_factorize_and_check_singular!(::Nothing, x) = _issingular(x), false
+
+# Needs Square Matrix
+"""
+    needs_square_A(alg)
+
+Returns `true` if the algorithm requires a square matrix.
+"""
+needs_square_A(::Nothing) = false
```
Comment (ChrisRackauckas) [inline src/utils.jl:266]: Should be in LinearSolve?

File: src/utils.jl [hunk 6]
```diff
@@ -206,8 +205,81 @@ function __get_concrete_algorithm(alg, prob)
...
+        return false
+    catch err
+        return true
+    end
+end
+for alg in (:QRFactorization, :FastQRFactorization, NormalCholeskyFactorization,
+    NormalBunchKaufmanFactorization)
+    @eval needs_square_A(::$(alg)) = false
+end
+for kralg in (LinearSolve.Krylov.lsmr!, LinearSolve.Krylov.craigmr!)
```
Comment (ChrisRackauckas) [inline src/utils.jl:283]: GMRES

Comment (avik-pal) [reply src/utils.jl:283]: GMRES requires square matrix

File: src/utils.jl (hunk 5 above)
Comment (avik-pal) [reply src/utils.jl:266]: Yes. Once all the tests pass I will upstream

File: src/gaussnewton.jl [hunk 7]
```diff
@@ -81,15 +81,29 @@ function SciMLBase.__init(prob::NonlinearLeastSquaresProblem{uType, iip}, alg_::
     kwargs...) where {uType, iip}
     alg = get_concrete_algorithm(alg_, prob)
     @unpack f, u0, p = prob
+
+    if !needs_square_A(alg.linsolve) && !(u isa Number) && !(u isa StaticArray)
```
Comment (axla-io) [inline src/gaussnewton.jl:85]: Should this be `u0` instead of `u`?

Comment (avik-pal) [reply src/gaussnewton.jl]: yes

File: src/gaussnewton.jl [hunk 8]
```diff
@@ -81,15 +81,29 @@ function SciMLBase.__init(prob::NonlinearLeastSquaresProblem{uType, iip}, alg_::
     kwargs...) where {uType, iip}
     alg = get_concrete_algorithm(alg_, prob)
     @unpack f, u0, p = prob
+
+    if !needs_square_A(alg.linsolve) && !(u isa Number) && !(u0 isa StaticArray)
```
Comment (axla-io) [inline src/gaussnewton.jl:85]: Should this be `u0` instead of `u`? I get a `u not defined` error

File: src/gaussnewton.jl [hunk 9]
```diff
@@ -99,12 +113,20 @@ end
...
-    linres = dolinsolve(alg.precs, linsolve; A = __maybe_symmetric(JᵀJ), b = _vec(Jᵀf),
-        linu = _vec(du), p, reltol = cache.abstol)
+    if JᵀJ !== nothing
+        __matmul!(JᵀJ, J', J)
+        __matmul!(Jᵀf, J', fu1)
+    end
+
+    # u = u - JᵀJ \ Jᵀfu
+    if cache.JᵀJ === nothing
+        linres = dolinsolve(alg.precs, linsolve; A = J, b = _vec(fu1), linu = _vec(du),
```
Comment (axla-io) [inline src/gaussnewton.jl:124]: I get an error stemming from here because the Jacobian is not square

Comment (avik-pal) [reply src/gaussnewton.jl:124]: what is the `linsolve.alg`? Is `A` or `b` some non-standard array type?

Comment (avik-pal) [reply src/gaussnewton.jl:124]: Also does specifying `linsolve = QRFactorization()` fix it?

Comment (axla-io) [reply src/gaussnewton.jl:124]: `linsolve.alg` is `nothing`. Changing to `QRFactorization()` gives a `not implemented `error. There could be some weird types yes, I'll try purging them...

Comment (axla-io) [reply src/gaussnewton.jl:124]: Still doesn't work with only standard arrays

Comment (avik-pal) [reply src/gaussnewton.jl:124]: Can you post the code for that?

File: src/utils.jl (hunk 6 above)
Comment (ChrisRackauckas) [reply src/utils.jl:286]: That's interesting 😅 . Algorithmically that's not required. It is an algorithm that provably gives the L2 solution on non-square matrices, just like QR.

File: src/raphson.jl [hunk 10]
```diff
@@ -96,7 +97,7 @@ function perform_step!(cache::NewtonRaphsonCache{true})
 
     # Line Search
     α = perform_linesearch!(cache.lscache, u, du)
-    @. u = u - α * du
+    axpy!(-α, du, u)
```
Comment (ChrisRackauckas) [inline src/raphson.jl:100]: does this have appropriate generic fallbacks?

Comment (avik-pal) [reply src/raphson.jl]: Done

**Comment by ChrisRackauckas** [comment]
This needs a rebase?

**Comment by avik-pal** [comment]
I am really unsure what to do with the Broyden/Klement failures. They fail on 1.9 but magically pass on 1.10

**Pull request merged**